File: 0684-redundant-connection/0684-redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        parent = [0] * (n+1)
        for i in range(len(parent)):
            parent[i] = i;
        def getParent(index):
            if (parent[index] == index):
                return index
            parent[index] = getParent(parent[index])
            return parent[index]
        def unionMethod(a, b):
            pa = getParent(a)
            pb = getParent(b)
            if (pa == pb):
                return False
            elif (pa < pb):
                parent[pb] = pa
            else:
                parent[pa] = pb
            return True
        
        for edge in edges:
            if (not unionMethod(edge[0], edge[1])):
                return edge
            
        return []
        

File: 0684-redundant-connection/0684-redundant-connection.cpp
class Solution {
public:
    
    vector<int> parent;
    int n;
    int getParent(int index) {
        if (index == parent[index]) return index;
        return getParent(parent[index]);
    }
    bool unionMethod(int a, int b) {
        int pa = getParent(a);
        int pb = getParent(b);
        if (pa == pb) return true;
        else if (pa < pb) {
            parent[pb] = pa;
        }
        else {
            parent[pa] = pb;
        }
        return false;
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        n = edges.size();
        parent = vector<int> (n+1);
        for (int i = 1; i <= n; ++i) parent[i] = i;
        vector<int> res;
        for (int i = 0; i < edges.size(); ++i) {
            if (unionMethod(edges[i][0], edges[i][1])) {
                res = edges[i];
            }
        }
        return res;
    }
};